File: hellogo/pkg/chess/optimized/intlist_test.go
package optimized

import (
	"testing"
)

func TestEmptyList(t *testing.T) {
	list := NewIntList()
	if list.Size() != 0 {
		t.Errorf("Expected size 0, got %d", list.Size())
	}
}

func TestAddAndGet(t *testing.T) {
	list := NewIntList()
	list.Add(10)
	list.Add(20)
	list.Add(30)

	if list.Size() != 3 {
		t.Errorf("Expected size 3, got %d", list.Size())
	}

	tests := []struct {
		index    int
		expected int
	}{
		{0, 10},
		{1, 20},
		{2, 30},
	}

	for _, test := range tests {
		if got := list.Get(test.index); got != test.expected {
			t.Errorf("list.Get(%d) = %d, want %d", test.index, got, test.expected)
		}
	}
}

func TestRemoveLast(t *testing.T) {
	list := NewIntList()
	list.Add(10)
	list.Add(20)


	if last := list.RemoveLast(); last != 20 {
		t.Errorf("Expected last element 20, got %d", last)
	}
	if list.Size() != 1 {
		t.Errorf("Expected size 1, got %d", list.Size())
	}

	if last := list.RemoveLast(); last != 10 {
		t.Errorf("Expected last element 10, got %d", last)
	}
	if list.Size() != 0 {
		t.Errorf("Expected size 0, got %d", list.Size())
	}
}

func TestContains(t *testing.T) {
	list := NewIntList()
	if list.Contains(10) {
		t.Error("Empty list should not contain any elements")
	}

	list.Add(10)
	list.Add(20)
	list.Add(30)


	tests := []struct {
		value    int
		expected bool
	}{
		{10, true},
		{20, true},
		{30, true},
		{40, false},
	}

	for _, test := range tests {
		if got := list.Contains(test.value); got != test.expected {
			t.Errorf("list.Contains(%d) = %v, want %v", test.value, got, test.expected)
		}
	}

	// Test after removing elements
	list.RemoveLast() // Removes 30
	if list.Contains(30) {
		t.Error("List should not contain 30 after removal")
	}

	list.Clear()
	if list.Contains(10) || list.Contains(20) {
		t.Error("List should be empty after clear")
	}
}

func TestClear(t *testing.T) {
	list := NewIntList()
	list.Add(10)
	list.Add(20)

	list.Clear()

	if list.Size() != 0 {
		t.Errorf("Expected size 0 after clear, got %d", list.Size())
	}
}

func TestCopyConstructor(t *testing.T) {
	// Test with empty list
	emptyList := NewIntList()
	emptyCopy := NewIntListFrom(emptyList)
	if emptyCopy.Size() != 0 {
		t.Errorf("Expected size 0, got %d", emptyCopy.Size())
	}

	// Test with non-empty list
	list := NewIntList()
	list.Add(10)
	list.Add(20)
	list.Add(30)


	copiedList := NewIntListFrom(list)

	// Verify the copy has the same elements
	if list.Size() != copiedList.Size() {
		t.Errorf("Expected size %d, got %d", list.Size(), copiedList.Size())
	}

	for i := 0; i < list.Size(); i++ {
		if list.Get(i) != copiedList.Get(i) {
			t.Errorf("Element at index %d differs: %d vs %d", i, list.Get(i), copiedList.Get(i))
		}
	}

	// Verify modifying the original doesn't affect the copy
	copySize := copiedList.Size()
	list.Add(40)
	if copiedList.Size() != copySize {
		t.Error("Modifying original affected copy's size")
	}
	if copiedList.Contains(40) {
		t.Error("Modifying original affected copy's contents")
	}

	// Verify modifying the copy doesn't affect the original
	copiedList.RemoveLast()
	if list.Size() == copiedList.Size() {
		t.Error("Modifying copy affected original's size")
	}
	if !list.Contains(30) {
		t.Error("Modifying copy affected original's contents")
	}
}

func TestCopyConstructorNil(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Error("Expected panic with nil source")
		}
	}()

	_ = NewIntListFrom(nil)
}

func TestAddAll(t *testing.T) {
	// Test adding to empty list
	source := NewIntList()
	source.Add(1)
	source.Add(2)
	source.Add(3)

	target := NewIntList()
	target.AddAll(source)

	if target.Size() != 3 {
		t.Fatalf("Expected size 3, got %d", target.Size())
	}

	expected := []int{1, 2, 3}
	for i, v := range expected {
		if target.Get(i) != v {
			t.Errorf("Expected %d at index %d, got %d", v, i, target.Get(i))
		}
	}

	// Test adding to non-empty list
	additional := NewIntList()
	additional.Add(4)
	additional.Add(5)

	target.AddAll(additional)


	if target.Size() != 5 {
		t.Fatalf("Expected size 5, got %d", target.Size())
	}

	expected = []int{1, 2, 3, 4, 5}
	for i, v := range expected {
		if target.Get(i) != v {
			t.Errorf("Expected %d at index %d, got %d", v, i, target.Get(i))
		}
	}

	// Test adding empty list
	prevSize := target.Size()
	target.AddAll(NewIntList())
	if target.Size() != prevSize {
		t.Errorf("Expected size %d after adding empty list, got %d", prevSize, target.Size())
	}
}

func TestLargeLists(t *testing.T) {
	const size1 = 1000
	const size2 = 2000

	// Test AddAll with large lists
	source := NewIntList()
	for i := 0; i < size1; i++ {
		source.Add(i)
	}

	target := NewIntList()
	for i := 0; i < size2; i++ {
		target.Add(i * 2)
	}

	target.AddAll(source)

	// Verify the result
	if target.Size() != size1+size2 {
		t.Fatalf("Expected size %d, got %d", size1+size2, target.Size())
	}

	// Check first size2 elements (original target elements)
	for i := 0; i < size2; i++ {
		expected := i * 2
		if got := target.Get(i); got != expected {
			t.Fatalf("At index %d: expected %d, got %d", i, expected, got)
		}
	}

	// Check remaining elements (from source)
	for i := 0; i < size1; i++ {
		expected := i
		if got := target.Get(size2 + i); got != expected {
			t.Fatalf("At index %d: expected %d, got %d", size2+i, expected, got)
		}
	}

	// Test Contains with large list
	// The target list contains:
	// 1. Numbers from 0 to (size2-1)*2 (even numbers)
	// 2. Then numbers from 0 to size1-1
	// So all numbers from 0 to size1-1 should be in the list (from the source)
	// And all even numbers up to (size2-1)*2 should be in the list (from the target)
	for i := 0; i < size1; i++ {
		if !target.Contains(i) {
			t.Errorf("Expected to find %d in the list", i)
		}
	}

	// Check some even numbers from the target
	for i := 0; i < 100; i += 2 {
		if !target.Contains(i) {
			t.Errorf("Expected to find %d in the list", i)
		}
	}

	// Also check some numbers that shouldn't be in the list
	if target.Contains(-1) {
		t.Error("Did not expect to find -1 in the list")
	}
	// size2*2 would be 4000, but since we only added numbers up to (size2-1)*2 (3998) and size1-1 (999)
	// 3999 is not in the list (it's odd and > size1-1)
	if target.Contains(3999) {
		t.Error("Did not expect to find 3999 in the list")
	}
	// Check for a number that's definitely not in the list
	// The list contains numbers up to max(1998, 999) = 1998
	// So 1999 is not in the list (it's odd and > 999)
	if target.Contains(1999) {
		t.Error("Did not expect to find 1999 in the list")
	}
}

File: hellojava/src/test/java/com/fathzer/chess/optimized/IntListTest.java
package com.fathzer.chess.optimized;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class IntListTest {
    private IntList list;

    @BeforeEach
    void setUp() {
        list = new IntList();
    }

    @Test
    void testEmptyList() {
        assertEquals(0, list.size());
    }

    @Test
    void testAddAndGet() {
        list.add(10);
        list.add(20);
        list.add(30);
        
        assertEquals(3, list.size());
        assertEquals(10, list.get(0));
        assertEquals(20, list.get(1));
        assertEquals(30, list.get(2));
    }

    @Test
    void testRemoveLast() {
        list.add(10);
        list.add(20);
        
        assertEquals(20, list.removeLast());
        assertEquals(1, list.size());
        assertEquals(10, list.get(0));
        
        assertEquals(10, list.removeLast());
        assertEquals(0, list.size());
    }
    
    @Test
    void testContains() {
        assertFalse(list.contains(10));
        
        list.add(10);
        list.add(20);
        list.add(30);
        
        assertTrue(list.contains(10));
        assertTrue(list.contains(20));
        assertTrue(list.contains(30));
        assertFalse(list.contains(40));
        
        // Test after removing elements
        list.removeLast(); // Removes 30
        assertTrue(list.contains(10));
        assertTrue(list.contains(20));
        assertFalse(list.contains(30));
        
        list.clear();
        assertFalse(list.contains(10));
        assertFalse(list.contains(20));
    }
    
    @ParameterizedTest
    @ValueSource(ints = {1, 10, 100, 1000})
    void testContainsWithLargeList(int size) {
        // Fill the list with even numbers
        for (int i = 0; i < size; i++) {
            list.add(i * 2);
        }
        
        // Test that all even numbers are contained
        for (int i = 0; i < size; i++) {
            if (i % 2 == 0) {
                assertTrue(list.contains(i), "Should contain " + i);
            } else {
                assertFalse(list.contains(i), "Should not contain " + i);
            }
        }
    }

    @Test
    void testClear() {
        list.add(10);
        list.add(20);
        
        list.clear();
        
        assertEquals(0, list.size());
    }
    
    @Test
    void testCopyConstructor() {
        // Test with empty list
        IntList emptyCopy = new IntList(list);
        assertEquals(0, emptyCopy.size());
        
        // Test with non-empty list
        list.add(10);
        list.add(20);
        list.add(30);
        
        IntList copy = new IntList(list);
        
        // Verify the copy has the same elements
        assertEquals(list.size(), copy.size());
        for (int i = 0; i < list.size(); i++) {
            assertEquals(list.get(i), copy.get(i), "Element at index " + i + " differs");
        }
        
        // Verify modifying the original doesn't affect the copy
        list.add(40);
        assertNotEquals(list.size(), copy.size());
        assertFalse(copy.contains(40));
        
        // Verify modifying the copy doesn't affect the original
        copy.removeLast();
        assertNotEquals(list.size() - 1, copy.size());
        assertTrue(list.contains(30));
    }
    
    @Test
    void testCopyConstructorNull() {
        assertThrows(NullPointerException.class, () -> new IntList(null));
    }
    
    @Test
    void testCopyConstructorLargeList() {
        // Test with a list larger than DEFAULT_CAPACITY
        final int size = 1000;
        for (int i = 0; i < size; i++) {
            list.add(i);
        }
        
        IntList copy = new IntList(list);
        assertEquals(size, copy.size());
        for (int i = 0; i < size; i++) {
            assertEquals(i, copy.get(i), "Element at index " + i + " differs");
        }
    }

    @Test
    void testAddAll() {
        // Test adding to empty list
        IntList other = new IntList();
        other.add(1);
        other.add(2);
        other.add(3);
        
        list.addAll(other);
        
        assertEquals(3, list.size());
        assertEquals(1, list.get(0));
        assertEquals(2, list.get(1));
        assertEquals(3, list.get(2));
        
        // Test adding to non-empty list
        IntList another = new IntList();
        another.add(4);
        another.add(5);
        
        list.addAll(another);
        
        assertEquals(5, list.size());
        assertEquals(4, list.get(3));
        assertEquals(5, list.get(4));
        
        // Test adding empty list
        int prevSize = list.size();
        list.addAll(new IntList());
        assertEquals(prevSize, list.size());
    }
    
    @Test
    void testAddAllWithLargeLists() {
        // Test adding lists that require capacity expansion
        final int size1 = 1000;
        final int size2 = 2000;
        
        // Fill the source list
        IntList source = new IntList();
        for (int i = 0; i < size1; i++) {
            source.add(i);
        }
        
        // Fill the target list
        for (int i = 0; i < size2; i++) {
            list.add(i * 2);
        }
        
        // Add all elements from source to target
        list.addAll(source);
        
        // Verify the result
        assertEquals(size1 + size2, list.size());
        
        // Check original elements are still there
        for (int i = 0; i < size2; i++) {
            assertEquals(i * 2, list.get(i));
        }
        
        // Check added elements are correct
        for (int i = 0; i < size1; i++) {
            assertEquals(i, list.get(size2 + i));
        }
    }
    
    @Test
    void testAddAllNull() {
        assertThrows(NullPointerException.class, () -> list.addAll(null));
    }

    @Test
    void testGetOutOfBounds() {
        list.add(10);
        assertThrows(IndexOutOfBoundsException.class, () -> list.get(-1));
    }

    @Test
    void testAutoResize() {
        // Add more elements than initial capacity (10)
        for (int i = 0; i < 20; i++) {
            list.add(i);
        }
        
        assertEquals(20, list.size());
        for (int i = 0; i < 20; i++) {
            assertEquals(i, list.get(i));
        }
    }
}
